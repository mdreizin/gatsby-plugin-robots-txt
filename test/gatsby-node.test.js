import { fs as memoryFs } from 'memfs';
import path from 'path';

jest.doMock('fs', jest.fn(() => memoryFs));

const fs = require('fs');
const { onPostBuild } = require('../src/gatsby-node');

const publicPath = './public';
const graphqlOptions = {
  site: {
    siteMetadata: {
      siteUrl: 'https://www.test.com'
    }
  }
};

function contentPath(filename) {
  return path.resolve(path.join(publicPath, filename));
}

function readContent(filename) {
  return fs.readFileSync(contentPath(filename)).toString();
}

describe('onPostBuild', () => {
  beforeAll(() => fs.mkdirpSync(path.resolve(publicPath)));

  it('should generate `robots.txt` using options', async () => {
    const output = './robots.txt';

    await onPostBuild(
      {
        graphql() {
          return Promise.resolve({ data: {} });
        }
      },
      {
        host: 'https://www.test.com',
        sitemap: 'https://www.test.com/sitemap.xml',
        output
      }
    );

    expect(readContent(output)).toMatchSnapshot();
  });

  it('should generate `robots.txt` using `graphql` options', async () => {
    const output = './robots-graphql.txt';

    await onPostBuild(
      {
        graphql() {
          return Promise.resolve({ data: graphqlOptions });
        }
      },
      {
        output
      }
    );

    expect(readContent(output)).toMatchSnapshot();
  });

  it('should not generate `robots.txt` in case of `graphql` errors', async () => {
    const output = './robots-graphql-err.txt';

    await expect(
      onPostBuild(
        {
          graphql() {
            return Promise.resolve({ errors: ['error1', 'error2'] });
          }
        },
        { output }
      )
    ).rejects.toEqual(new Error('error1, error2'));

    expect(fs.existsSync(contentPath(output))).toBeFalsy();
  });

  it('should not generate `robots.txt` in case of I/O errors', async () => {
    const output = './robots-io-err.txt';

    const spy = jest
      .spyOn(fs, 'writeFile')
      .mockImplementation((file, data, callback) =>
        callback(new Error('error'))
      );

    await expect(
      onPostBuild(
        {
          graphql() {
            return Promise.resolve({ data: graphqlOptions });
          }
        },
        { output }
      )
    ).rejects.toEqual(new Error('error'));

    expect(fs.existsSync(contentPath(output))).toBeFalsy();

    spy.mockRestore();
  });

  it('should generate `robots.txt` using `env` options', async () => {
    const output = './robots-env.txt';

    await onPostBuild(
      {
        graphql() {
          return Promise.resolve({ data: {} });
        }
      },
      {
        host: 'https://www.test.com',
        sitemap: 'https://www.test.com/sitemap.xml',
        output,
        env: {
          test: {
            policy: [{ userAgent: '*', disallow: ['/'] }]
          }
        }
      }
    );

    expect(readContent(output)).toMatchSnapshot();
  });

  it('should generate `robots.txt` using `env` options and `resolveEnv` function', async () => {
    const output = './robots-env-custom.txt';

    await onPostBuild(
      {
        graphql() {
          return Promise.resolve({ data: {} });
        }
      },
      {
        host: 'https://www.test.com',
        sitemap: 'https://www.test.com/sitemap.xml',
        output,
        resolveEnv: () => 'custom',
        env: { custom: { policy: [{ userAgent: '*', disallow: ['/'] }] } }
      }
    );

    expect(readContent(output)).toMatchSnapshot();
  });
});
