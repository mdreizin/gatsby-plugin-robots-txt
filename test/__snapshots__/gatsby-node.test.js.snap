// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onPostBuild should generate \`robots.txt\` using \`env\` options 1`] = `
"User-agent: *
Disallow: /
Sitemap: https://www.test.com/sitemap.xml
Host: https://www.test.com
"
`;

exports[`onPostBuild should generate \`robots.txt\` using \`env\` options and \`resolveEnv\` function 1`] = `
"User-agent: *
Disallow: /
Sitemap: https://www.test.com/sitemap.xml
Host: https://www.test.com
"
`;

exports[`onPostBuild should generate \`robots.txt\` using \`graphql\` options 1`] = `
"User-agent: *
Allow: /
Sitemap: https://www.test.com/sitemap.xml
Host: https://www.test.com
"
`;

exports[`onPostBuild should generate \`robots.txt\` using options 1`] = `
"User-agent: *
Allow: /
Sitemap: https://www.test.com/sitemap.xml
Host: https://www.test.com
"
`;
